### PR TITLE
#618 Add validate model method

### PR DESCRIPTION
**Pull request by 3rdvision**
| - | - |
| --- | --- |
| Issue | https://github.com/ripe-tech/ripe-white/issues/618 |
| Decisions | Add model validation method to validate model on purchase button. |

### CODE REVIEW

File: src/js/api/brand.js [hunk 1]
```diff
@@ -491,6 +491,96 @@ ripe.Ripe.prototype.onInitialsP = function(options) {
...
+ *  - 'brand' - The name of the brand to be considerd when validating
+ *  - 'model' - The name of the model to be considerd when validating
+ *  - 'version' - The target build version to be considerd when validating
+ *  - 'parts' - The parts to be validated
+ *  - 'engraving' - The engraving value to be validated
+ *  - 'size' - The size to be validated
+ * @param {Function} callback Function with the result of the request.
+ * @returns {XMLHttpRequest} Resulting information for the callback execution.
+ */
+ripe.Ripe.prototype.validateModel = function(options, callback) {
```
Comment (gcandal) [inline src/js/api/brand.js:508]: Move this after `getFactory` (same order as ripe-core/src/ripe_core/controllers/api/brand.py)

Comment (gcandal) [inline src/js/api/brand.js:508]: `getValidateModel` since it's a `GET`

File: src/js/api/brand.js [hunk 2]
```diff
@@ -491,6 +491,96 @@ ripe.Ripe.prototype.onInitialsP = function(options) {
     });
 };
 
+/**
+ * Returns the result of the server side validation of model configurations
+ * considering a given build's brand, model and version.
+ *
+ * @param {Object} options An object with options, such as:
```
Comment (gcandal) [inline src/js/api/brand.js:498]: Some stuff missing:

```
        version = self.field("version", None)
        variant = self.field("variant", None)
        parts = self.field("p", [], cast = list)
        initials = self.field("initials", None)
        engraving = self.field("engraving", None)
        initials_extra = self.field("initials_extra", None, multiple = True)
        gender = self.field("gender", None)
        size = self.field("size", None)
```

Comment (gcandal) [reply src/js/api/brand.js:498]: at least gender and variant

**Review by gcandal — CHANGES_REQUESTED** [review]

File: src/js/api/brand.js (hunk 2 above)
Comment (3rdvision) [reply src/js/api/brand.js:498]: I believe everything is here
gender: https://github.com/ripe-tech/ripe-sdk/blob/316b55f4ba8d49a422470a3db660e0f3c2cd72d5/src/js/api/brand.js#L547
variant: https://github.com/ripe-tech/ripe-sdk/blob/316b55f4ba8d49a422470a3db660e0f3c2cd72d5/src/js/api/brand.js#L541

They are presented in the same order as in ripe-core

Comment (gcandal) [reply src/js/api/brand.js]: Yep, they're just missing in the javadoc.

File: src/js/base/api.js [hunk 3]
```diff
@@ -1166,8 +1166,8 @@ ripe.Ripe.prototype._partsToPartsM = function(parts) {
     return partsM;
 };
 
-ripe.Ripe.prototype._partsMToQuery = function(partsM, sort = true) {
-    const queryL = [];
+ripe.Ripe.prototype._partsMToTriplets = function(partsM, sort = true) {
```
Comment (joamag) [inline src/js/base/api.js:1169]: Good idea (method re-usage), we need to back-port this solution to the Python version

File: src/js/api/brand.js [hunk 4]
```diff
@@ -491,6 +534,54 @@ ripe.Ripe.prototype.onInitialsP = function(options) {
...
+    const engraving = options.engraving === undefined ? this.engraving : options.engraving;
+    const initialsExtra =
+        options.initialsExtra === undefined ? this.initialsExtra : options.initials_extra;
+    const gender = options.gender === undefined ? this.gender : options.gender;
+    const size = options.size === undefined ? this.size : options.size;
+    const brand = options.brand === undefined ? this.brand : options.brand;
+    const model = options.model === undefined ? this.model : options.model;
+    const url = `${this.url}brands/${brand}/models/${model}/validate`;
+    const params = {};
+    if (version !== undefined && version !== null) {
```
Comment (joamag) [inline src/js/api/brand.js:554]: I believe we should re-use this kind of code from (`_getConfigInfoOptions`)

```
    if (queryOptions) options = this._getQueryOptions(options);
    if (initialsOptions) options = this._getInitialsOptions(options);
```

File: src/js/api/brand.js [hunk 5]
```diff
@@ -286,6 +286,49 @@ ripe.Ripe.prototype.getFactoryP = function(options) {
...
+ *  - 'brand' - The name of the brand to be considerd when validating
+ *  - 'model' - The name of the model to be considerd when validating
+ *  - 'version' - The target build version to be considerd when validating
+ *  - 'parts' - The parts to be validated
+ *  - 'engraving' - The engraving value to be validated
+ *  - 'size' - The size to be validated
+ * @param {Function} callback Function with the result of the request.
+ * @returns {XMLHttpRequest} Resulting information for the callback execution.
+ */
+ripe.Ripe.prototype.getValidateModel = function(options, callback) {
```
Comment (joamag) [inline src/js/api/brand.js:303]: maybe call this method only `validateModel`

**Review by joamag — CHANGES_REQUESTED** [review]

File: src/js/base/api.js (hunk 3 above)
Comment (3rdvision) [reply src/js/base/api.js:1169]: Same method re-usage for ripe-core :arrow_right: https://github.com/ripe-tech/ripe-core/pull/4500

**Review by joamag — APPROVED** [review]

**Pull request merged**